sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/ExpressionCodegen.scala

```diff
@@ -0,0 +1,248 @@
+  }
+
+  /**
+   * Given a child expression, retrieves previously evaluated columns referred by it or
+   * deferred expressions which are needed to evaluate it.
+   */
+  def getInputVars(ctx: CodegenContext, child: Expression): Seq[(Expression, ExprCode)] = {
+    if (ctx.currentVars == null) {
+      return Seq.empty
+    }
+
+    child.flatMap {
+      // An evaluated variable.
+      case b @ BoundReference(ordinal, _, _) if ctx.currentVars(ordinal) != null &&
+          ctx.currentVars(ordinal).code == "" =>
+        Seq((b, ctx.currentVars(ordinal)))
+
+      // An input variable which is not evaluated yet. Tracks down to find any evaluated variables
+      // in the expression path.
+      // E.g., if this expression is "d = c + 1" and "c" is not evaluated. We need to track to
+      // "c = a + b" and see if "a" and "b" are evaluated. If they are, we need to return them so
+      // to include them into parameters, if not, we track down further.
+      case BoundReference(ordinal, _, _) if ctx.currentVars(ordinal) != null =>
+        trackDownVar(ctx, ctx.currentVars(ordinal))
+
+      case _ => Seq.empty
+    }.distinct
+  }
+
+  /**
+   * Tracks down previously evaluated columns referred by the generated code snippet.
+   */
+  def trackDownVar(ctx: CodegenContext, exprCode: ExprCode): Seq[(Expression, ExprCode)] = {
+    val exprCodes = mutable.Queue[ExprCode](exprCode)
+    val inputVars = mutable.ArrayBuffer.empty[(Expression, ExprCode)]
+
+    while (exprCodes.nonEmpty) {
+      exprCodes.dequeue().inputVars.foreach { inputVar =>
+        if (inputVar.exprCode.code == "") {
+          inputVars += ((inputVar.expr, inputVar.exprCode))
+        } else {
+          exprCodes.enqueue(inputVar.exprCode)
+        }
+      }
+    }
+    inputVars
+  }
+
+  /**
+   * Helper function to calculate the size of an expression as function parameter.
+   */
+  def calculateParamLength(ctx: CodegenContext, input: Expression): Int = {
+    ctx.javaType(input.dataType) match {
+      case (ctx.JAVA_LONG | ctx.JAVA_DOUBLE) if !input.nullable => 2
+      case ctx.JAVA_LONG | ctx.JAVA_DOUBLE => 3
+      case _ if !input.nullable => 1
+      case _ => 2
+    }
+  }
+
+  /**
+   * In Java, a method descriptor is valid only if it represents method parameters with a total
+   * length of 255 or less. `this` contributes one unit and a parameter of type long or double
+   * contributes two units.
+   */
+  def getParamLength(
+      ctx: CodegenContext,
+      inputs: Seq[Expression],
+      subExprs: Seq[Expression]): Int = {
+    // Initial value is 1 for `this`.
+    1 + (inputs ++ subExprs).distinct.map(calculateParamLength(ctx, _)).sum
+  }
+
+  /**
+   * Given the lists of input attributes and variables to this expression, returns the strings of
+   * funtion parameters. The first is the variable names used to call the function, the second is
+   * the parameters used to declare the function in generated code.
+   */
+  def prepareFunctionParams(
```

Comment:
How about avoid duplication of similar logic in `getParamsForSubExprs`?  
In `getParamsForSubExprs`, I think that `subExprs` corresponds to `inputAttrs` and `inputVars` corresponds to `subExprs.map(ctx.subExprEliminationExprs(_))`.

Reply:
Yes, sure.